File: rca-android/app/src/main/java/com/quickseries/rca/repository/ModuleRepository.java
package com.quickseries.rca.repository;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.quickseries.rca.common.ApiResponse;
import com.quickseries.rca.common.NetworkBoundResource;
import com.quickseries.rca.common.Resource;
import com.quickseries.rca.controller.ExecutorController;
import com.quickseries.rca.local.ApplicationDatabase;
import com.quickseries.rca.local.ModuleEntity;
import com.quickseries.rca.remote.ApiService;
import com.quickseries.rca.service.NetworkConnectivityService;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.Executor;

import javax.inject.Inject;
import javax.inject.Singleton;

import retrofit2.Response;

/**
 * Repository that handles Module objects.
 */
@Singleton
public class ModuleRepository {

    private static final String TAG = ModuleRepository.class.getSimpleName();

    private ApiService apiService;
    private ApplicationDatabase applicationDatabase;
    private final ExecutorController  executorController;

    @Inject
    public ModuleRepository(ExecutorController executorController, ApplicationDatabase applicationDatabase, ApiService apiService) {
        this.apiService = apiService;
        this.applicationDatabase = applicationDatabase;
        this.executorController = executorController;
    }

    public LiveData<Resource<List<ModuleEntity>>> loadModules(String authorizationToken, String appId) {

        return new NetworkBoundResource<List<ModuleEntity>,List<ModuleEntity>>(executorController) {

            @Override
            protected void saveCallResult(@NonNull List<ModuleEntity> item) {
                applicationDatabase.moduleDao().insertAll(item);
            }

            @Override
            protected boolean shouldFetch(@Nullable List<ModuleEntity> data) {
                return data == null || data.isEmpty();
            }

            @NonNull
            @Override
            protected LiveData<List<ModuleEntity>> loadFromDb() {
                return applicationDatabase.moduleDao().loadAllModules();
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<ModuleEntity>>> createCall() {
                return apiService.fetchModules(authorizationToken, appId);
            }

        }.asLiveData();
    }
}

File: core/src/main/java/com/anou/prototype/core/repository/ModuleRepository.kt
package com.anou.prototype.core.repository

import androidx.lifecycle.LiveData
import com.anou.prototype.core.api.ApiService
import com.anou.prototype.core.db.ApplicationDatabase
import com.anou.prototype.core.db.ModuleEntity
import com.anou.prototype.core.service.NetworkConnectivityService
import com.anou.prototype.core.strategy.LocalDataAwareFirstStrategy
import com.anou.prototype.core.strategy.ResourceWrapper
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Deferred

class ModuleRepository(
        val applicationDatabase: ApplicationDatabase,
        val apiService: ApiService,
        private val networkConnectivityService: NetworkConnectivityService
) : BaseRepository() {

    fun loadModules(): LiveData<ResourceWrapper<List<ModuleEntity>>> = object : LocalDataAwareFirstStrategy<List<ModuleEntity>>() {
//        override fun isRemoteAvailable() = networkConnectivityService.getConnectionType() != NetworkConnectivityService.ConnectionType.TYPE_NO_INTERNET
        override fun isRemoteAvailable() = true

        override suspend fun fetchData(): Deferred<List<ModuleEntity>> {
            return apiService.fetchModules()
        }

        override suspend fun readData(): Deferred<LiveData<List<ModuleEntity>>> {
            return CompletableDeferred(applicationDatabase.moduleDao().loadAllModules())
        }

        override suspend fun writeData(data: List<ModuleEntity>) {
            applicationDatabase.moduleDao().insertAll(data)
        }

    }.asLiveData()
}